diem-move/diem-framework/experimental/sources/NFT.move

```diff
@@ -1,17 +1,30 @@
 address 0x1 {
     module NFT {
-        use Std::Errors;
         use Std::Event;
         use Std::GUID;
         use Std::Option::{Self, Option};
         use Std::Signer;
         use Std::Vector;
 
+        /// Struct representing a semi-fungible or non-fungible token (depending on the supply).
+        /// There can be multiple tokens with the same id (unless supply is 1). Each token's
+        /// corresponding token metadata is stored inside a TokenData inside TokenDataCollection
+        /// under the creator's address.
+        /// The TokenData might be inlined together with the token in case the token is unique, i.e., its balance is 1
+        /// (we might choose to extend inlining for the non-unique NFTs in future).
+        /// The TokenData can also be separated out to a separate creator's collection in order to normalize the
+        /// data layout: we'd want to keep a single instance of the token data in case its balance is large.
+        struct Token<TokenType: copy + store + drop> has key, store {
```

Comment:
maybe it's just me feeling paranoid, having copy + drop for tokens makes it smell fungible somehow lol

Reply:
Yea, well, the fungible part is not the token itself but the `TokenType`.
The only reason we need the `copy` part here is that we cannot return the reference, so in order to read the content of the token data, one needs to either "pull it out" (the hot potato strategy), or copy. The "pull it out" for reading is quite counter-intuitive in my opinion, so had to use copying. I agree that it's sad.

Reply:
If it's `drop` couldn't we completely drop the token from storage?

Reply:
It's a good point. The `drop` is on the `TokenType` (specifically, on token metadata).
The question is about the way we want to consume the token metadata, since Move doesn't allow returning reference to the metadata. There are two ways:
1. The "hot potato" style: we're pulling it out from storage upon read, and require the reader to return it back to storage when they're done. I find this. approach kind of confusing, with potentially poor perf (that part I don't know).
2. The "copy" style: someone is allowed to copy the metadata if they need to read it (it's also the approach taken by StarCoin). In case a reader retrieves a copy of metadata, we should probably allow the reader to drop it, hence `TokenType` now has both `copy` and `drop`. Note that it has nothing to do with `copy` and `drop` of the `Token` itself (which is not copyable or droppable).